ops-bedrock: Beacon-chain devnet with Dencun + Ecotone upgrade

sebastianst opened the pull request:
**Description**

This PR makes the local devnet to use a real lighthouse Beacon node.

It also enables Delta by default on all e2e tests, plus L1 enables Cancun after a few blocks, and Ecotone a shortly after.

In a follow up PR, Ecotone and Cancun will be enabled at genesis.

The PR also improves fork selection in individual tests and fixes from fork tests (e.g. one Ecotone test didn't call its activation function).

The PR introduces a new package `op-e2e/devnet` to run e2e tests against the local devnet. It provides a minimal `System` abstraction that works with the withdrawals test, which got extracted and made into a reusable test.

ℹ️ _For reviewers:_ the files in `ops-bedrock/data` are auto-generated.

**Tests**

Many fixed.

**Additional context**

Explicit fork selection in tests is somewhat improved, but is still brittle. We should completely remove fork selection by setting individual `<Fork>TimeOffset` fields in the deploy config and do it by a modular constructor of the configs using the `With...` option pattern.

Open TODOs are tracked at https://github.com/ethereum-optimism/optimism/issues/10968

**Metadata**

- Fixes https://github.com/ethereum-optimism/protocol-quest/issues/81

protolambda committed:
ops-bedrock: beacon-chain devnet with fork to L1 Dencun and L2 Ecotone

ops-bedrock: clean up file names, drop unused stateviz

ops-bedrock: dockerfile for lighthouse from source

ops-bedrock: devnet validator keystores

ops-bedrock: dev chain config

generate genesis.ssz, work in progress devnet setup

lighthouse CLI tweaks

ops-bedrock: more CLI tweaks

devnet: fix validator keys loading

ops-bedrock: cleanup L1 geth setup and jwt-secret usage

ops-bedrock: fix L1 engine API endpoint

op-chain-ops: insert dev L1 contracts

devnet fixes

more fixes

schedule L2 Delta and Ecotone forks

op-chain-ops: fix lint

ops-bedrock: docker-compose config comments about special options

ops-bedrock: adjust beacon node log level

Co-authored-by: Sebastian Stammler <seb@oplabs.co>

protolambda commented:
Squashed and rebased on latest `develop`, in case anyone wants to run this or experiment with Ecotone locally.

tynes commented:
Wondering if we plan on moving forward with this as we will be needing a devnet for interop development soon

protolambda commented:
@tynes yes, moving forward with this. Currently blocked on 2 things:
- L1/L2 genesis refactor completion
- eth2-testnet-genesis tool update, to generate a L1 dencun genesis, instead of a shapella one and having to schedule an upgrade. Working on updating this (implementing dencun support in underlying libs now).

protolambda commented:
`eth2-testnet-genesis` supports Dencun at genesis now

sebastianst committed:
devnet: Add eth2-testnet-genesis installation step

sebastianst committed:
op-e2e: fix pre-Ecotone tests

sebastianst committed:
ops-bedrock: Fix entrypoint scripts

sebastianst committed:
op-e2e: Fix TestPreEcotone

sebastianst committed:
op-e2e: Fix hardfork validation tests

sebastianst commented on ops-bedrock/docker-compose.yml:
```diff
@@ -27,52 +29,98 @@ services:
       args:
         OP_STACK_GO_BUILDER: us-docker.pkg.dev/oplabs-tools-artifacts/images/op-stack-go:devnet
     image: us-docker.pkg.dev/oplabs-tools-artifacts/images/op-node:devnet
+    # The --l1.beacon.fetch-all-sidecars is required until
+    # the lighthouse image includes the following API bugfix: https://github.com/sigp/lighthouse/pull/5118
     command: >
       op-node
       --l1=ws://l1:8546
+      --l1.beacon=http://l1-bn:5052
+      --l1.beacon.fetch-all-sidecars
```
we can remove `fetch-all-sidecars`

sebastianst committed:
op-e2e: Enable L1 Cancun at genesis for Ecotone system configs

sebastianst committed:
ci: use eth2-testnet-genesis tag v0.10.0

sebastianst committed:
op-e2e: use <fork>SystemConfig instead of manipulating fields

sebastianst committed:
op-e2e/actions: process beacon block root for Cancun L1s

sebastianst committed:
op-e2e/actions: fix TestBlockTimeBatchType

sebastianst committed:
op-e2e: add beacon endpoint to L1 setup

sebastianst committed:
op-e2e/actions: fix more tests...

sebastianst committed:
op-e2e/actions: fix manual setting of delta offset

sebastianst committed:
Merge branch 'develop' into ecotone-beacon-devnet

sebastianst committed:
op-e2e/actions: set blob source in plasma tests

sebastianst committed:
devnet: sleep at start of tests

sebastianst committed:
ops-bedrock: remove fetch-all-sidecars for l1 lh from docker-compose

protolambda commented on op-chain-ops/genesis/layer_one.go:
```diff
@@ -95,6 +97,17 @@ func BuildL1DeveloperGenesis(config *DeployConfig, dump *gstate.Dump, l1Deployme
+		return nil, fmt.Errorf("failed to insert beacon deposit contract into L1 dev genesis: %w", err)
+	}
+
+	// For 4788, make sure the 4788 beacon-roots contract is there.
+	// (required to be there before L1 Dencun activation)
+	memDB.CreateAccount(predeploys.EIP4788ContractAddr)
+	memDB.SetNonce(predeploys.EIP4788ContractAddr, 1)
+	memDB.SetCode(predeploys.EIP4788ContractAddr, predeploys.EIP4788ContractCode)
```
For consistency with actual L1 we should also set the nonce on the EIP4788 sender account, otherwise it could run in a weird failed-tx case where the destination contract address already exists.

sebastianst replied:
added, please check that it's correct

protolambda replied:
checked, matches EIP, thanks!

protolambda commented on op-e2e/Makefile:
```diff
@@ -6,7 +6,7 @@ ifdef JUNIT_FILE
 else
 	go_test = go test
-	go_test_flags = -v
+	go_test_flags = # -v
```
Why this change?

sebastianst replied:
thanks for spotting, shouldn't have been committed, artifact from local testing

sebastianst committed:
op-e2e: fix Makefile go test flags

sebastianst committed:
op-chain-ops: Add 4788 deployer address to L1 genesis creation

tynes commented on op-e2e/actions/reorg_test.go:
```diff
@@ -148,11 +148,11 @@ func ReorgFlipFlop(gt *testing.T, deltaTimeOffset *hexutil.Uint64) {
-		//require.NotEqual(t, verifier.L2Unsafe().Hash, ref.ParentHash, "TODO off by one, engine syncs A0 after reorging back from B, while rollup node only inserts up to A0 (excl.)")
-		//require.Equal(t, verifier.L2Unsafe(), ref, "verifier safe head of engine matches rollup client")
+		// ref, err := verifEngClient.L2BlockRefByLabel(t.Ctx(), eth.Unsafe)
+		// require.NoError(t, err)
```
This is not part of this PR but confused as to why this code is commented out in the first place

sebastianst replied:
My auto-save-formatter created this diff. No idea what's going on here. git blame shows that @protolambda added this TODO on 2022-10-22, so it's quite dated. Would prefer not to deal with this in this PR. I'll check with him if this can be deleted.

sebastianst replied:
Discussed internally with proto that we can delete this TODO code, the test has the right assertions and good coverage.

tynes commented on ops-bedrock/docker-compose.yml:
```diff
@@ -172,6 +217,9 @@ services:
       OP_BATCHER_PLASMA_DA_SERVER: "http://da-server:3100"
+      # uncomment to use blobs
+      # (requires L1 Dencun and L2 Ecotone activation first)
+      # OP_BATCHER_DATA_AVAILABILITY_TYPE: blobs
```
Seems like we should not have to make a diff to use this but instead use env var templating, similar to this line:

```
      OP_CHALLENGER_GAME_FACTORY_ADDRESS: ${DGF_ADDRESS}
```

so we could do something like:

```
      OP_BATCHER_DATA_AVAILABILITY_TYPE: ${OP_BATCHER_DATA_AVAILABILITY_TYPE:-calldata}
```

sebastianst replied:
Good idea, I will do this in the follow up to this PR where I enable L1 Cancun at genesis in the devnet. Right now, this cannot be set to blobs because L1 Cancun only activates after 0x30 seconds.

tynes commented on .circleci/config.yml:
```diff
@@ -1239,6 +1239,10 @@ jobs:
+      - run:
+          name: Install eth2-testnet-genesis
+          command: |
+            go install -v github.com/protolambda/eth2-testnet-genesis@v0.10.0
```
This doesn't use `versions.json` to ensure the correct version is used

sebastianst replied:
`versions.json` isn't used at all in the circleci config, see e.g. geth and nvm, which also have a hardcoded version. So I just followed this pattern. Would you still prefer to use the `versions.json` here?

sebastianst replied:
OK added this, it works. Added a follow up ticket (https://github.com/ethereum-optimism/optimism/issues/10398) to use `versions.json` more consistently in the cci config.

sebastianst committed:
op-e2e: remove outdated TODO code

sebastianst committed:
ci: read eth2-testnet-genesis version from json

sebastianst committed:
Merge branch 'develop' into ecotone-beacon-devnet

sebastianst committed:
Merge branch 'develop' into ecotone-beacon-devnet

tynes commented:
How is the progress on this coming along?

sebastianst commented:
> How is the progress on this coming along?

@tynes Blocked by a bunch of other more urgent work unfortunately. Hope to get to it on Friday or Monday.

sebastianst committed:
Merge branch 'develop' into ecotone-beacon-devnet

sebastianst committed:
fix l1 genesis & cci config

sebastianst committed:
rename beacon-data/deposit_contract_block.txt

sebastianst committed:
Merge remote-tracking branch 'origin/develop' into ecotone-beacon-devnet

sebastianst committed:
Merge remote-tracking branch 'origin/develop' into ecotone-beacon-devnet

sebastianst committed:
cci: dump l1-bn and l1-vs logs

sebastianst committed:
few improvements

sebastianst committed:
Move to op-e2e devnet withdrawals test

sebastianst committed:
Merge remote-tracking branch 'origin/develop' into ecotone-beacon-devnet

sebastianst commented on bedrock-devnet/devnet/__init__.py:
```diff
@@ -332,10 +336,10 @@ def devnet_test(paths):
-        CommandPreset('erc20-test',
-          ['npx', 'hardhat',  'deposit-erc20', '--network',  'devnetL1',
-           '--l1-contracts-json-path', paths.addresses_json_path, '--signer-index', '14'],
-          cwd=paths.tasks_dir, timeout=8*60),
+        # CommandPreset('erc20-test',
+        #   ['npx', 'hardhat',  'deposit-erc20', '--network',  'devnetL1',
+        #    '--l1-contracts-json-path', paths.addresses_json_path, '--signer-index', '14'],
+        #   cwd=paths.tasks_dir, timeout=8*60)
```
Didn't mean to commit this diff. If reviewers are happy with the op-e2e devnet test, I will just remove this altogether. Can then remove the `devnet-tasks` in a follow up.

tynes replied:
I would like to remove these devnet tasks, perhaps its a better idea to see them pass in CI against the new devnet before removing them

sebastianst replied:
@tynes They don't pass and I cannot get them to work. I wasted a lot of time on this. This is the reason I switched to Go tests.

tynes commented on op-e2e/e2eutils/secrets.go:
```diff
@@ -189,3 +192,14 @@ func (a *Addresses) All() []common.Address {
 	}
 }
+
+func (s *Secrets) AccountAtIdx(idx int) *ecdsa.PrivateKey {
```
perhaps `uint` because negative values don't make sense here?

sebastianst replied:
It's common to use `int`s in Go for indices that don't make sense to be negative. So I just followed that convention. E.g. see stdlib, it uses `int` indices everywhere. The user of this API is usually an expert who wouldn't put in negative numbers. So I'd just leave it as is.

tynes commented:
We are moving towards removing the L2OO and being a fault proofs only sort of system, just wanna make sure that we are on the same page and that this takes that into account, ie no deep assumptions about L2OO in this diff

sebastianst commented:
> We are moving towards removing the L2OO and being a fault proofs only sort of system, just wanna make sure that we are on the same page and that this takes that into account, ie no deep assumptions about L2OO in this diff

@tynes yes, this PR just adds lighthouse, the default local devnet is still FP enabled.

sebastianst committed:
set l1 block time to 6s in devnet template

sebastianst committed:
Merge branch 'develop' into ecotone-beacon-devnet

protolambda committed:
fix game resolve helper, to handle resolve race condition

sebastianst committed:
lower da windows

sebastianst committed:
Merge remote-tracking branch 'origin/develop' into ecotone-beacon-devnet

sebastianst committed:
devnet: bump plasma windows to 16

tchardin committed:
fix: adjust plasma test params for smaller challenge window sizes

protolambda committed:
Merge pull request #11240 from latticexyz/tc/fix-plasma-e2e

fix: adjust plasma test params for smaller challenge window sizes

sebastianst committed:
Merge remote-tracking branch 'origin/develop' into ecotone-beacon-devnet

sebastianst committed:
bump lighthouse, remove src Dockerfile

sebastianst committed:
revert devnet python script commenting out

the test isn't run from this script any more anyways
that was just temporary while debugging

sebastianst committed:
Merge remote-tracking branch 'origin/develop' into ecotone-beacon-devnet

sebastianst committed:
remove deprecated options

ligthouse deprecated http-allow-sync-stalled
docker-compose deprecated the version tag

sebastianst committed:
retry start block fetching in L2 genesis cmd

sebastianst committed:
add CL to devnet L1 startup step

sebastianst committed:
ignore receipt for resolveClaim call in FinalizeWithdrawal

protolambda commented:
Quick update about CI; I believe @sebastianst has a draft of invoking the transaction-trace on failure, to capture the `output`, and determine the error code, to fix the FP part of the withdrawal devnet tests.

Since the challenger runs in parallel to the Go test, the resolve-steps by the Go test don't always turn into success, as the claim, and then the game, may already be resolved.

sebastianst committed:
forward tx trace as custom error type, use in withdrawal test

sebastianst committed:
return receipt in SendTx

zhiqiangxu commented:
> This PR makes the local devnet to use a real lighthouse Beacon node.

I guess this feature is optional, right ?
